AH-152: DOI Form filling

prdanelli opened the pull request.

cjcolvar commented on app/controllers/concerns/hyku_addons/doi_controller_behavior.rb:
```diff
@@ -0,0 +1,68 @@
+          data: formatted_work,
+          curation_concern: params[:curation_concern],
+          fields: {
+            json: curation_concern_class.json_fields,
+            date: curation_concern_class.date_fields,
```
I think it would be best not to require the JS to know about these.  I think it might tie the JS too closely to the implementation details on the backend.  How about splitting these fields on the server-side so the JSON response is just the fields that match up with the form fields on the page?

cjcolvar commented on app/controllers/concerns/hyku_addons/doi_controller_behavior.rb:
```diff
@@ -0,0 +1,68 @@
+        curation_concern.constantize
+      end
+
+      def curation_concern_reader_class
```
Should this be a writer instead of a reader?

prdanelli replied:
@cjcolvar would you be able to clarify this one for me?

cjcolvar commented on lib/hyku_addons/engine.rb:
```diff
@@ -230,6 +230,7 @@ def self.dynamically_include_mixins
       User.include HykuAddons::UserEmailFormat
       Bolognese::Writers::RisWriter.include Bolognese::Writers::RisWriterBehavior
       Hyrax::GenericWorksController.include HykuAddons::WorksControllerBehavior
+      Hyrax::DOI::HyraxDOIController.include HykuAddons::DOIControllerBehavior
```
If we come up with a new pattern that we like and is generic, I think it would be good to push it back upstream to `hyrax-doi` and hopefully remove this override eventually.

cjcolvar reviewed (commented):
The JS looks good at a quick glance.  We can chat tomorrow if any of my comments don't make sense. :slightly_smiling_face:

cjcolvar commented on app/models/concerns/hyku_addons/generic_work_overrides.rb:
```diff
@@ -124,7 +124,7 @@ module GenericWorkOverrides
 
       self.json_fields += %i[editor current_he_institution]
-      self.date_fields += %i[event_date related_exhibition_date]
+      self.date_fields += %i[event_date related_exhibition_date date_created]
```
I don't believe that `date_created` is a field on our models. (It might be in upstream GenericWork but not on ours.)

prdanelli replied:
You are correct. I added it as a test because `hyrax_work` is only giving me `date_created` in the json... and I forgot to remove it.

cjcolvar reviewed (commented):
I think a lot of this work should be pushed into a Bolognese writer module that creates work objects using our work types instead of upstream Hyku's GenericWork.  If it creates a work object that matches the requested `model_class` then the JSON fields should only need to be parsed from the string form and included in the response.

prdanelli commented on app/controllers/concerns/hyku_addons/doi_controller_behavior.rb:
```diff
@@ -0,0 +1,65 @@
+          input = reader_class.new(input: doi)
+
+          # Use that response to build the object as we would expect it
+          meta = reader_class.new(input: input.string)
```
@cjcolvar There might be a better way of doing this, if so, could you give me your thoughts?

The DOI request returns a partially formed reader object, with `string` being assigned the result of the response. However the readers all assume that `meta` is available and so the above gets around that by rebuilding the reader with the XML string.

prdanelli commented:
@cjcolvar this is ready to review now. I came to the conclusion that we don't really need to know which fields are on the form/model. Everything seems to be just fine without passing `creation_concern` as the `work_type`, so i've removed the extra complexity. 

If you think we might need to know this in the future, I would always roll back the commit in which i removed it.

cjcolvar reviewed (approved):
:tada: Let's get this deployed! :shipit:
